### PR TITLE
Add snapshot test suite to tfp-automation

### PR DESCRIPTION
**Pull request by markusewalker**
### PR Description
This PR adds support for the `snapshot` test suite to the `tfp-automation` framework. In short, the following functionality is being added:
- RKE1/RKE2/K3S cluster configuration for etcd backups (local && S3)
- Creating etcd snapshots and restoring snapshots for RKE1/RKE2/K3S
- Test files for restore etcd only, kubernetes version and etcd, cluster configuration and kubernetes version and etcd
- Polished up misc. files for better code readability

### CODE REVIEW

File: config/config.go
Comment (slickwarren) [inline config/config.go:149]: nothing else has 'omitempty' should it be rm'd from etcd variables?

File: framework/set/setRKE1Config.go
Comment (slickwarren) [inline framework/set/setRKE1Config.go:79]: would it be worth setting default consts for these hard coded ints and strings? are these just what we have by default in rancher, and you have to manually set it in tf else they are set to nil?

File: framework/set/setETCDSnapshots.go
Comment (slickwarren) [inline framework/set/setETCDSnapshots.go:9]: can you modify this comment to specify that it is being set in main.tf, not actually creating the resource in this function?

File: framework/set/setETCDSnapshots.go
Comment (slickwarren) [inline framework/set/setETCDSnapshots.go:23]: can you modify this comment to specify that it is being set in main.tf, not actually restoring the resource in this function?

File: framework/set/setRKE2K3sConfig.go
Comment (slickwarren) [inline framework/set/setRKE2K3sConfig.go:109]: same comment for defaulting these strings.

File: framework/set/setRKE2K3sConfig.go
Comment (slickwarren) [inline framework/set/setRKE2K3sConfig.go:119]: maybe because omitempty is enabled, but does this not give an error if terraformConfig.ETCD is nil? might be fine as is since ETCD is a pointer, but just checking if that was tested.

File: config/config.go
Comment (slickwarren) [inline config/config.go:141]: seeing you break this out in set/setProviders.go made me realize, this should be broken out as well. In the config, we should have `aws`, `azure`, `linode`, etc. provider structs. Similar to how we do it in rancher/rancher. That way, the main terraformConfig isn't bloated with each individual provider's variables. i.e. vsphere will be gross as there are a ton of unique variables for it. 

could you make a new task to track this? Probably too much to be included in this PR.

File: tests/provisioning/README.md
Comment (slickwarren) [inline tests/provisioning/README.md:32]: is k3s2 or k3s1 available?

File: tests/snapshot/snapshot_restore_k8s_upgrade_test.go
Comment (slickwarren) [inline tests/snapshot/snapshot_restore_k8s_upgrade_test.go:103]: node roles missing from test name

File: tests/snapshot/snapshot_restore_k8s_upgrade_test.go
Comment (slickwarren) [inline tests/snapshot/snapshot_restore_k8s_upgrade_test.go:88]: nit; is it necessary to define a new config.TerratestConfig object here, or could you just define the config.Snapshots instead?

File: tests/snapshot/snapshot.go
Comment (slickwarren) [inline tests/snapshot/snapshot.go:64]: super nit; easier to read if you set a default value for this i.e. `false` and only have an if statement instead of if/else.

File: tests/snapshot/snapshot.go
Comment (slickwarren) [inline tests/snapshot/snapshot.go:128]: super nit; might make more sense to break this out into `snapshotRKE1` and `restoreRKE1` and equivalent for rke2/k3s so that this is more modular for future test cases.

File: tests/snapshot/snapshot_restore_upgrade_strategy_test.go
Comment (slickwarren) [inline tests/snapshot/snapshot_restore_upgrade_strategy_test.go:92]: continue super nit

File: tests/upgrading/README.md
Comment (slickwarren) [inline tests/upgrading/README.md:34]: is it k3s2 or k3s1 for this version?

**Review by slickwarren — COMMENTED** [review]
lgtm, a few clarifying points and nits. 
waiting for freeform jobs for final approval

File: tests/snapshot/snapshot.go
Comment (markusewalker) [reply tests/snapshot/snapshot.go:128]: Yeah the majority of this is just copy/paste from `rancher/shepherd`. If it is alright with you, let's make the change there first and update here? Just for parity so it's not confusing for the time being.

File: config/config.go
Comment (markusewalker) [reply config/config.go:141]: Great idea. I'll see how much it is to break it down here. If it's too much, I'll create a new task, but if it's not that big of a deal, I'll just update it in this PR.

File: framework/set/setRKE2K3sConfig.go
Comment (markusewalker) [reply framework/set/setRKE2K3sConfig.go:119]: Yeah I tested that. It doesn't give an error, in fact, I had to have it this way because if you try to have S3 but not the additional check for it, that is what gives an error.

File: tests/snapshot/snapshot_restore_k8s_upgrade_test.go
Comment (markusewalker) [reply tests/snapshot/snapshot_restore_k8s_upgrade_test.go:88]: Yeah, it has to be here. I tried doing the latter first, but from how I refactored the config, you have to define the `config.TerratestConfig` first.

File: tests/snapshot/snapshot.go
Comment (markusewalker) [reply tests/snapshot/snapshot.go:64]: I see what you're saying here. Would it be too much trouble if we make the change in rancher/shepherd first and then here? Just to have them all on parity.

File: framework/set/setRKE1Config.go
Comment (markusewalker) [reply framework/set/setRKE1Config.go:79]: This is an oversight, I actually meant to have these be dynamically set. I'll change that.

File: framework/set/setRKE2K3sConfig.go
Comment (markusewalker) [reply framework/set/setRKE2K3sConfig.go:109]: These are just the default values set in Rancher. They aren't defined values to be dynamically set like in the prior comment, so I left it as is.

File: config/config.go
Comment (markusewalker) [reply config/config.go]: Resolved

File: framework/set/setRKE1Config.go
Comment (markusewalker) [reply framework/set/setRKE1Config.go]: Resolved

File: framework/set/setETCDSnapshots.go
Comment (markusewalker) [reply framework/set/setETCDSnapshots.go]: Resolved

File: config/config.go
Comment (markusewalker) [reply config/config.go]: Resolved

File: tests/snapshot/snapshot_restore_k8s_upgrade_test.go
Comment (markusewalker) [reply tests/snapshot/snapshot_restore_k8s_upgrade_test.go]: Resolved

File: tests/upgrading/README.md
Comment (markusewalker) [reply tests/upgrading/README.md:34]: When I was testing, I grabbed a version available found in Rancher and this was the version.

File: config/config.go
Comment (slickwarren) [inline config/config.go:130]: hey, can you update all these vars to `omitempty` so that they aren't required?

Comment (markusewalker) [reply config/config.go]: Resolved

File: config/aws_config.go
Comment (slickwarren) [inline config/aws_config.go:15]: I would think that some of these aren't required / could be null and therefore could have omitempty? Not entirely sure though and will leave this up to you

**Review by slickwarren — APPROVED** [review]
LGTM, one small question but not blocking approval

File: config/aws_config.go
Comment (markusewalker) [reply config/aws_config.go]: Resolved. I honestly forgot to update this lol

**Pull request merged**
